examples/erc20/tests/erc20.rs

```diff
@@ -1216,3 +1256,237 @@ async fn error_when_transfer_from(alice: Account, bob: Account) -> Result<()> {
+
+    assert_eq!(initial_nonce + uint!(1_U256), nonce);
+
+    Ok(())
+}
+
+#[e2e::test]
+async fn permit_rejects_reused_signature(
+    alice: Account,
+    bob: Account,
+) -> Result<()> {
+    let contract_addr = deploy(alice.url(), &alice.pk(), None).await?;
+    let contract_alice = Erc20::new(contract_addr, &alice.wallet);
+    let alice_addr = alice.address();
+    let bob_addr = bob.address();
+
+    let balance = uint!(10_U256);
+    let _ = watch!(contract_alice.mint(alice_addr, balance))?;
+
+    let struct_hash = prepare_struct_hash(
+        alice_addr,
+        bob_addr,
+        balance,
+        U256::ZERO,
+        FAIR_DEADLINE,
+    );
+
+    let typed_data_hash =
+        prepare_typed_data_hash(domain_separator!(contract_alice), struct_hash);
+    let signature = alice
+        .sign_hash(&alloy::primitives::B256::from_slice(
+            typed_data_hash.as_slice(),
+        ))
+        .await;
+
+    let _ = watch!(contract_alice.permit(
+        alice_addr,
+        bob_addr,
+        balance,
+        FAIR_DEADLINE,
+        signature.v().y_parity_byte_non_eip155().unwrap(),
+        signature.r().into(),
+        signature.s().into()
+    ))?;
+
+    let err = send!(contract_alice.permit(
+        alice_addr,
+        bob_addr,
+        balance,
+        FAIR_DEADLINE,
+        signature.v().y_parity_byte_non_eip155().unwrap(),
+        signature.r().into(),
+        signature.s().into()
+    ))
+    .expect_err("should return `ERC2612InvalidSigner`");
+
+    let struct_hash = prepare_struct_hash(
+        alice_addr,
+        bob_addr,
+        balance,
+        U256::from(1),
+        FAIR_DEADLINE,
+    );
+
+    let typed_data_hash =
+        prepare_typed_data_hash(domain_separator!(contract_alice), struct_hash);
+
+    let recovered = signature
+        .recover_address_from_prehash(&alloy::primitives::B256::from_slice(
+            typed_data_hash.as_slice(),
+        ))
+        .expect("should recover");
+
+    assert!(err.reverted_with(Erc20::ERC2612InvalidSigner {
+        signer: recovered,
+        owner: alice_addr
+    }));
+
+    Ok(())
+}
+
+#[e2e::test]
+async fn permit_rejects_invalid_signature(
+    alice: Account,
+    bob: Account,
+) -> Result<()> {
+    let contract_addr = deploy(alice.url(), &alice.pk(), None).await?;
+    let contract_alice = Erc20::new(contract_addr, &alice.wallet);
+    let alice_addr = alice.address();
+    let bob_addr = bob.address();
+
+    let balance = uint!(10_U256);
+    let _ = watch!(contract_alice.mint(alice_addr, balance))?;
+
+    let struct_hash = prepare_struct_hash(
+        alice_addr,
+        bob_addr,
+        balance,
+        U256::ZERO,
+        FAIR_DEADLINE,
+    );
+
+    let typed_data_hash =
+        prepare_typed_data_hash(domain_separator!(contract_alice), struct_hash);
+    let signature = bob
+        .sign_hash(&alloy::primitives::B256::from_slice(
+            typed_data_hash.as_slice(),
+        ))
```

Comment:
Ah, this is probably because the primitives version that `alloy` uses is not the same as the `alloy_primitives` we use. Let's create an issue to fix this.

Reply:
Unfortunately we need to wait for #179 to have same `alloy` version...

Reply:
yeah, 😢